File: app/src/main/java/miui/statusbar/lyric/view/LyricTextView.java
package miui.statusbar.lyric.view;

/*
 * Copyright (C) 2019 The Android Open Source Project
 *               2020 The exTHmUI Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.widget.TextView;

public class LyricTextView extends TextView {

    private boolean isStop = true;
    private float textLength = 0f;
    private float viewWidth = 0f;
    private float speed = 4f;
    private float x = 0f;
    private String text;

    public static final int startScrollDelay = 500;
    public static final int invalidateDelay = 10;

    private final Paint mPaint;

    Runnable mStartScrollRunnable = this::startScroll;

    public LyricTextView(Context context) {
        this(context, null);
    }

    public LyricTextView(Context context, AttributeSet attrs) {
        this(context, attrs, 0x01010084);
    }

    public LyricTextView(Context context, AttributeSet attrs, int defStyleAttr) {
        this(context, attrs, defStyleAttr, 0);
    }

    public LyricTextView(Context context, AttributeSet attrs, int defStyleAttr,
                         int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        mPaint = getPaint();
    }

    private void init() {
        x = 0;
        textLength = getTextLength();
        viewWidth = getWidth();
    }

    @Override
    protected void onDetachedFromWindow() {
        removeCallbacks(mStartScrollRunnable);
        super.onDetachedFromWindow();
    }

    @Override
    protected void onTextChanged(CharSequence text, int start, int lengthBefore, int lengthAfter) {
        super.onTextChanged(text, start, lengthBefore, lengthAfter);
        stopScroll();
        this.text = text.toString();
        init();
        postInvalidate();
        postDelayed(mStartScrollRunnable, startScrollDelay);
    }

    @Override
    public void setTextColor(int color) {
        if (mPaint != null) mPaint.setColor(color);
        postInvalidate();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        if (canvas != null && text != null) {
            float y = getHeight() / 2 + Math.abs(mPaint.ascent() + mPaint.descent()) / 2;
            canvas.drawText(text, x, y, mPaint);
        }
        if (!isStop) {
            if (viewWidth - x + speed >= textLength) {
                x = viewWidth > textLength ? 0 : viewWidth - textLength;
                stopScroll();
            } else {
                x -= speed;
            }
            invalidateAfter(invalidateDelay);
        }
    }

    private void invalidateAfter(long delay) {
        removeCallbacks(invalidateRunnable);
        postDelayed(invalidateRunnable, delay);
    }

    private final Runnable invalidateRunnable = this::invalidate;

    public void startScroll() {
        init();
        isStop = false;
        postInvalidate();
    }

    public void stopScroll() {
        isStop = true;
        removeCallbacks(mStartScrollRunnable);
        postInvalidate();
    }

    private float getTextLength() {
        return mPaint == null ? 0 : mPaint.measureText(text);
    }

    public void setSpeed(float speed) {
        this.speed = speed;
    }
}

File: app/src/main/java/statusbar/lyric/view/LyricTextView.kt
/*
 * StatusBarLyric
 * Copyright (C) 2021-2022 fkj@fkj233.cn
 * https://github.com/577fkj/StatusBarLyric
 *
 * This software is free opensource software: you can redistribute it
 * and/or modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either
 * version 3 of the License, or any later version and our eula as published
 * by 577fkj.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * and eula along with this software.  If not, see
 * <https://www.gnu.org/licenses/>
 * <https://github.com/577fkj/StatusBarLyric/blob/main/LICENSE>.
 */

package statusbar.lyric.view

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Shader
import android.widget.TextView
import statusbar.lyric.config.XposedOwnSP.config

class LyricTextView(context: Context) : TextView(context) {
    private var isScrolling = false
    private var textLength = 0f
    private var viewWidth = 0f
    private var iconWidth = 0f
    private var scrollSpeed = 4f
    private var currentX = 0f
    private val iconSwitch = config.iconSwitch
    private val lyricStartMargins = config.lyricStartMargins
    private val lyricEndMargins = config.lyricEndMargins
    private val iconStartMargins = config.iconStartMargins
    private val startScrollRunnable = Runnable { updateScrollPosition() }

    init {
        paint.style = Paint.Style.FILL_AND_STROKE
    }

    override fun onDetachedFromWindow() {
        stopScroll()
        super.onDetachedFromWindow()
    }

    override fun onTextChanged(text: CharSequence, start: Int, lengthBefore: Int, lengthAfter: Int) {
        super.onTextChanged(text, start, lengthBefore, lengthAfter)
        textLength = getTextLength()
        currentX = 0f
        startScroll()
    }

    override fun setTextColor(color: Int) {
        paint.color = color
        postInvalidate()
    }

    fun setLinearGradient(shader: Shader) {
        paint.shader = shader
        postInvalidate()
    }

    fun setStrokeWidth(width: Float) {
        paint.strokeWidth = width
        postInvalidate()
    }

    override fun onDraw(canvas: Canvas) {
        val y = (height - (paint.descent() + paint.ascent())) / 2
        text?.let { canvas.drawText(it.toString(), currentX, y, paint) }
    }

    private fun updateScrollPosition() {
        val realTextLength = textLength + lyricEndMargins + lyricStartMargins
        val realLyricWidth = viewWidth - if (iconSwitch) iconWidth + iconStartMargins else 0f
        if (realTextLength <= realLyricWidth) {
            currentX = 0f
            stopScroll()
        } else if (realLyricWidth - currentX >= realTextLength) {
            currentX = realLyricWidth - realTextLength
            stopScroll()
        } else {
            currentX -= scrollSpeed
        }
    }

    override fun computeScroll() {
        if (isScrolling) {
            postDelayed(startScrollRunnable, 1000)
            postInvalidate()
        }
    }

    private fun startScroll() {
        isScrolling = true
    }

    private fun stopScroll() {
        isScrolling = false
        removeCallbacks(startScrollRunnable)
    }

    private fun getTextLength(): Float {
        return text?.let { paint.measureText(it.toString()) } ?: 0f
    }

    fun setScrollSpeed(speed: Float) {
        this.scrollSpeed = speed
    }

    fun maxViewWidth(float: Float) {
        viewWidth = float
    }

    fun iconWidth(width: Float) {
        if (config.iconSwitch) {
            iconWidth = width
        }
    }
}